Sources/XCTest/XCTestCaseType.swift

```diff
@@ -0,0 +1,119 @@
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2016 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See http://swift.org/LICENSE.txt for license information
+// See http://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//
+//  XCTestCaseType.swift
+//  Base protocol (and extension with default methods) for test cases
+//
+
+public typealias TestMethod = XCTestCaseType throws -> Void
+
+public protocol XCTestCaseType {
+    init()
+
+    func setUp()
+    func tearDown()
+
+    // In the Objective-C version of XCTest, it is possible to discover all tests when the test is executed by asking the runtime for all methods and looking for the string "test". In Swift, we ask test providers to tell us the list of tests by implementing this property.
+    static var allTests: [(String, TestMethod)] { get }
+}
+
+/// Wrapper function allowing a static reference to a test method to fit 
+/// the signature required by XCTestCaseType.allTests
+public func test<T: XCTestCaseType>(testFunc: T -> () throws -> Void) -> TestMethod {
+    return { testCaseType in
+        guard let testCase: T = testCaseType as? T else {
+            fatalError("Attempt to invoke test on class \(T.self) with incompatible instance type \(testCaseType.dynamicType)")
+        }
+
+        try testFunc(testCase)()
+    }
+}
```

Comment:
Totally agreed about documentation being critical here. I was just holding off on fleshing that out until a consensus was reached on whether this API direction seems reasonable.